src/lex.cpp

```diff
@@ -1288,6 +1292,52 @@ void lexer::parse_modern_octal_number() {
       diag_unexpected_characters_in_octal_number>(input);
 }
 
+void lexer::check_precision_loss(const char8* number_begin,
+                                 const char8* input) {
+  string8_view number_literal(number_begin,
+                              static_cast<size_t>(input - number_begin));
+  const size_t GUARANTEED_ACC_LENGTH = 15;
+  const size_t MAX_ACC_LENGTH = 309;
+  if (number_literal.size() > GUARANTEED_ACC_LENGTH) {
+    std::string cleaned_string = "";
+    for (char8 c : number_literal) {
+      if (c != '_') {
+        cleaned_string.push_back(static_cast<char>(c));
+      }
+    }
+    if (cleaned_string.size() > GUARANTEED_ACC_LENGTH) {
+      if (cleaned_string.size() > MAX_ACC_LENGTH) {
+        this->diag_reporter_->report(diag_number_literal_will_lose_precision{
+            .characters = source_code_span(number_begin, input),
+            .rounded_val = u8"inf"sv,
+        });
+        return;
+      }
+      double num = strtod(cleaned_string.c_str(), nullptr);
+      std::array<char, MAX_ACC_LENGTH + 1> result_string;
+      int rc = std::snprintf(result_string.data(), result_string.size(), "%.0f",
+                             num);
+      QLJS_ALWAYS_ASSERT(rc >= 0 &&
+                         static_cast<size_t>(rc) < result_string.size());
```

Comment:
Nit: Split into two assert calls:
```suggestion
      QLJS_ALWAYS_ASSERT(rc >= 0);
      QLJS_ALWAYS_ASSERT(static_cast<size_t>(rc) < result_string.size());
```